File: src/api/analytics.ts
import {
	type FileSessionsResponse,
	file_sessions_response_schema,
} from "@contracts/analytics/file-sessions";
import {
	type FileSizeResult,
	file_size_result_schema,
	type ProjectFileStats,
	project_file_stats_schema,
} from "@contracts/analytics/files";
import {
	type AnalyticsOverview,
	analytics_overview_schema,
} from "@contracts/analytics/overview";
import {
	type TimeBreakdown,
	time_breakdown_schema,
} from "@contracts/analytics/time";
import {
	type ToolDetailResponse,
	tool_detail_response_schema,
} from "@contracts/analytics/tools";
import { session_entries_response_schema } from "@contracts/sessions/replay";
import {
	type SessionSummary,
	session_summary_schema,
} from "@contracts/sessions/summary";
import {
	type ProjectSummary,
	project_summary_schema,
} from "@contracts/shared/primitives";
import { z } from "zod";
import type { SessionEntriesResponse } from "@/components/session-viewer/types";
import { commands } from "@/platform/ipc";

export async function list_projects(): Promise<ProjectSummary[]> {
	const raw = await commands.analytics.list_projects();
	return z.array(project_summary_schema).parse(raw);
}

export async function get_analytics_overview(
	project_path?: string,
	range_days?: number,
): Promise<AnalyticsOverview> {
	const raw = await commands.analytics.get_analytics_overview({
		projectPath: project_path ?? null,
		rangeDays: range_days ?? null,
	});
	return analytics_overview_schema.parse(raw);
}

export async function get_session_detail(
	session_id: string,
): Promise<SessionSummary> {
	const raw = await commands.analytics.get_session_detail({
		sessionId: session_id,
	});
	return session_summary_schema.parse(raw);
}

export async function get_all_sessions(
	project_path?: string,
	range_days?: number,
): Promise<SessionSummary[]> {
	const raw = await commands.analytics.get_all_sessions({
		projectPath: project_path ?? null,
		rangeDays: range_days ?? null,
	});
	return z.array(session_summary_schema).parse(raw);
}

export async function resync_sessions(): Promise<AnalyticsOverview> {
	const raw = await commands.analytics.resync_sessions();
	return analytics_overview_schema.parse(raw);
}

export async function get_project_file_stats(
	project_path: string,
	range_days?: number,
): Promise<ProjectFileStats> {
	const raw = await commands.analytics.get_project_file_stats({
		projectPath: project_path,
		rangeDays: range_days ?? null,
	});
	return project_file_stats_schema.parse(raw);
}

export async function get_time_breakdown(
	range_days: number,
	project_path?: string,
): Promise<TimeBreakdown> {
	const raw = await commands.analytics.get_time_breakdown({
		rangeDays: range_days,
		projectPath: project_path ?? null,
	});
	return time_breakdown_schema.parse(raw);
}

export async function get_session_entries(
	session_id: string,
): Promise<SessionEntriesResponse> {
	const raw = await commands.analytics.get_session_entries({
		sessionId: session_id,
	});
	// Validate with strict discriminated-union schema, then return as
	// component-level types (structurally identical, downstream narrows by entry.type)
	return session_entries_response_schema.parse(raw) as SessionEntriesResponse;
}

export async function get_tool_details(
	tool_name: string,
	project_path?: string,
	range_days?: number,
): Promise<ToolDetailResponse> {
	const raw = await commands.analytics.get_tool_details({
		toolName: tool_name,
		projectPath: project_path ?? null,
		rangeDays: range_days ?? null,
	});
	return tool_detail_response_schema.parse(raw);
}

export async function get_file_sizes(
	paths: string[],
): Promise<FileSizeResult[]> {
	const raw = await commands.analytics.get_file_sizes({ paths });
	return z.array(file_size_result_schema).parse(raw);
}

export async function get_sessions_for_files(
	project_path: string,
	file_paths: string[],
	range_days?: number,
): Promise<FileSessionsResponse> {
	const raw = await commands.analytics.get_sessions_for_files({
		projectPath: project_path,
		filePaths: file_paths,
		rangeDays: range_days ?? null,
	});
	return file_sessions_response_schema.parse(raw);
}

File: src-tauri/src/commands/analytics.rs
use tauri::State;

use crate::models::analytics::{AnalyticsOverview, ToolDetailResponse, ProjectFileStats, TimeBreakdown, ProjectSummary, FileSizeResult};
use crate::models::session::{SessionSummary, SessionEntriesResponse};
use crate::cache::SessionCache;
use crate::qmd_log_cache::QmdLogCache;

#[tauri::command]
pub async fn get_analytics_overview(cache: State<'_, SessionCache>, project_path: Option<String>, range_days: Option<u32>) -> Result<AnalyticsOverview, String> {
    cache.get_analytics_overview(project_path.as_deref(), range_days.unwrap_or(0)).await
}

#[tauri::command]
pub async fn list_projects(cache: State<'_, SessionCache>) -> Result<Vec<ProjectSummary>, String> {
    cache.list_projects().await
}

#[tauri::command]
pub async fn get_session_detail(cache: State<'_, SessionCache>, session_id: String) -> Result<SessionSummary, String> {
    cache.get_session_detail(&session_id).await
}

#[tauri::command]
pub async fn get_all_sessions(
    cache: State<'_, SessionCache>,
    project_path: Option<String>,
    range_days: Option<u32>,
) -> Result<Vec<SessionSummary>, String> {
    cache.get_all_sessions(project_path.as_deref(), range_days.unwrap_or(0)).await
}

#[tauri::command]
pub async fn resync_sessions(
    cache: State<'_, SessionCache>,
    qmd_log_cache: State<'_, QmdLogCache>,
) -> Result<AnalyticsOverview, String> {
    // Force resync the session cache
    cache.resync().await?;
    // Invalidate QMD log cache so it rebuilds lazily on next request
    qmd_log_cache.invalidate().await;
    // Return fresh analytics overview (all time, all projects)
    cache.get_analytics_overview(None, 0).await
}

#[tauri::command]
pub async fn get_project_file_stats(
    cache: State<'_, SessionCache>,
    project_path: String,
    range_days: Option<u32>,
) -> Result<ProjectFileStats, String> {
    cache.get_project_file_stats(&project_path, range_days.unwrap_or(0)).await
}

#[tauri::command]
pub async fn get_time_breakdown(cache: State<'_, SessionCache>, range_days: u32, project_path: Option<String>) -> Result<TimeBreakdown, String> {
    cache.get_time_breakdown(range_days, project_path.as_deref()).await
}

#[tauri::command]
pub async fn get_tool_details(
    cache: State<'_, SessionCache>,
    tool_name: String,
    project_path: Option<String>,
    range_days: Option<u32>,
) -> Result<ToolDetailResponse, String> {
    cache.get_tool_details(&tool_name, project_path.as_deref(), range_days.unwrap_or(0)).await
}

#[tauri::command]
pub async fn get_file_sizes(
    paths: Vec<String>,
) -> Result<Vec<FileSizeResult>, String> {
    Ok(SessionCache::get_file_sizes(paths))
}

#[tauri::command]
pub async fn get_session_entries(
    cache: State<'_, SessionCache>,
    session_id: String,
) -> Result<SessionEntriesResponse, String> {
    cache.get_session_entries(&session_id).await
}
